pkg/apis/events/v1/conversion.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	v1 "k8s.io/api/events/v1"
+	conversion "k8s.io/apimachinery/pkg/conversion"
+	k8s_api "k8s.io/kubernetes/pkg/apis/core"
+	k8s_api_v1 "k8s.io/kubernetes/pkg/apis/core/v1"
+)
+
+func Convert_v1_Event_To_core_Event(in *v1.Event, out *k8s_api.Event, s conversion.Scope) error {
+	if err := autoConvert_v1_Event_To_core_Event(in, out, s); err != nil {
+		return err
+	}
+	if err := k8s_api_v1.Convert_v1_ObjectReference_To_core_ObjectReference(&in.Regarding, &out.InvolvedObject, s); err != nil {
+		return err
+	}
+	if err := k8s_api_v1.Convert_v1_EventSource_To_core_EventSource(&in.DeprecatedSource, &out.Source, s); err != nil {
```

Comment:
we might want to be converting `reportingController` to Source, see: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/events/event_recorder.go#L88 cc @wojtek-t

Reply:
I added a conversion for that and removed the todo in event_recorder file. Let me know if there's anything else I should change.

Reply:
This is not what @yastij was referring for - what he meant was not "DeprecatedSource" byt "reportingController".

So IIUC what we should is:
- check if DeprecatedSource is false or "DeprecatedSource == ReportingController" and then convert.

But what if those are not equal? I'm not entirely sure we can do that safely now.

Reply:
@wojtek-t is right. Ideally we need to codify the intent in the conversion so it's even enforced through custom clients. what we could do is something like:

- from v1 -> core/v1, if DeprecatedSource != ReportingController, ReportingController is preferred, because if it came through a `core/v1` conversion we are pretty sure that they are equal
- from core/v1 -> v1, ensure we populate DeprecatedSource and ReportingController with the value provided in the `Source` field

Reply:
Sorry I don't fully get what should be changed here, especially from v1 -> corev1. From what I understand, If `in.DeprecatedSource.Component == in.ReportingController`, we convert `in.DeprecatedSource` to `out.Source`; otherwise, we assign `in.ReportingController` to `out.Source.Component`? If this is correct, I don't see the how that different from what I did here.
For corev1 -> v1, it looks like what I did is correct?

Reply:
Yes

Reply:
This won't pass the TestRoundTripTypes test.
In the test, both Source.Component and ReportingController of `core/v1` event are set but with different values. After the roundtrip, the original ReportingController will be forgotten. Do we still want to do the conversion? If yes, did I miss something?

Reply:
@liggitt @wojtek-t - can we customize fuzz testing for this ? If that's something not supported today, then I'm not sure If worth doing

Reply:
i would say let's ignore it and do the same we did for v1beta1: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/events/v1beta1/conversion.go

Reply:
that's fair

Reply:
Cool, I've removed these conversions.